---
title: "Genotype inference"
execute:
  echo: false
  warning: false
  message: false
format: html
---

```{r}
source("functions.R")
```

```{r}
library(tidyverse)
library(reshape2)
library(ggplot2)
library(ggtext)
library(vdjbasevis)
```

```{r}
num_subjects = length(unique(data_$sample))

```

```{r,echo=FALSE}

imgt_genes = 'IGHV'

genes <- unique(bag_data[bag_data$in_baseline_reference=="TRUE","gene"])
ASCs_on_ref= genes[grepl(imgt_genes, genes)]
lenght_ASCs_on_ref=length(ASCs_on_ref)

genes_on_genotype <- unique(bag_data[bag_data$sample_count > 0,"gene"])
ASCs_on_genotype = genes_on_genotype[grepl(imgt_genes, genes_on_genotype)]
novel_ASCs_on_genotype = setdiff(ASCs_on_ref,ASCs_on_genotype)
length_novel_ASCs_on_genotype = length(novel_ASCs_on_genotype)
lenght_ASCs_on_genotype = length(setdiff(ASCs_on_genotype,novel_ASCs_on_genotype) )

lenght_ASCs_not_represented = lenght_ASCs_on_ref - lenght_ASCs_on_genotype

allele <- unique(bag_data[bag_data$in_baseline_reference=="TRUE","allele"])
alleles_on_ref=allele[grepl(imgt_genes, allele)]
lenght_alleles_on_ref=length(alleles_on_ref)

allele_on_genotype <- unique(bag_data[bag_data$sample_count > 0,"allele"])
alleles_on_genotype = allele_on_genotype[grepl(imgt_genes,allele_on_genotype)]
novel_alleles_on_genotype = setdiff(c(alleles_on_genotype),alleles_on_ref)
length_novel_alleles_on_genotype = length(novel_alleles_on_genotype)
lenght_alleles_on_genotype = length(setdiff(c(alleles_on_genotype),novel_alleles_on_genotype) )

lenght_alleles_not_represented = lenght_alleles_on_ref - lenght_alleles_on_genotype
```

For the genotype inference we applied a pre-genotype filter that required sequences to have no mutations within the V region, a single V allele assignment, and alignment starting from position one of the V germline. Only alleles having a threshold z Score above 0 were retained. For IGHD genotype inference, the selection was further refined by including only sequences with no mutations in the D region and a single assigned D allele. Additionally, only alleles that exceeded the z-score threshold above 4 were retained. For each allele, the median ratio of sequence length observed in the repertoire to the allele length was calculated using sequences annotated with that specific IGHD allele from subjects carrying the allele in their genotype, considering only alleles where the ratio was above 0.5.

The Piglet allele-based method was used for genotype inference based on AIRR-seq data. It compares allele usage against a population-derived threshold to determine allele inclusion in the genotype. A z score is calculated for each allele, measuring its significance in the repertoire. The formula incorporates allele count, frequency threshold, and repertoire depth, with higher z scores indicating a stronger likelihood of allele presence in the genotype.

In a detailed analysis, it was found that of the `r lenght_ASCs_on_ref` existing IGHV clusters, `r lenght_ASCs_not_represented` are not represented in the genotypes of the `r num_subjects` subjects analyzed. `r lenght_alleles_not_represented` out of `r lenght_alleles_on_ref` existing IGHV alleles were not found in the genotypes of the subjects.

There are `r length_novel_alleles_on_genotype` IGHV novel allele that were found in the genotypes of these `r num_subjects` subjects.